thor/observation_source.py

```diff
@@ -0,0 +1,132 @@
+import abc
+from typing import TypeAlias
+
+import numpy as np
+import quivr as qv
+from adam_core.observations import detections, exposures
+
+from . import orbit
+
+
+class Observations:
+    """Observations represents a collection of exposures and the
+    detections they contain.
+
+    The detections may be a filtered subset of the detections in the
+    exposures.
+    """
+
+    detections: detections.PointSourceDetections
+    exposures: exposures.Exposures
+    linkage: qv.Linkage[detections.PointSourceDetections, exposures.Exposures]
+
+    def __init__(
+        self,
+        detections: detections.PointSourceDetections,
+        exposures: exposures.Exposures,
+    ):
+        self.detections = detections
+        self.exposures = exposures
+        self.linkage = qv.Linkage(
+            detections,
+            exposures,
+            left_keys=detections.exposure_id,
+            right_keys=exposures.id,
+        )
+
+
+class ObservationSource(abc.ABC):
+    """An ObservationSource is a source of observations for a given test orbit.
+
+    It has one method, gather_observations, which takes a test orbit
+    and returns a collection of Observations.
+
+    """
+
+    @abc.abstractmethod
+    def gather_observations(self, test_orbit: orbit.TestOrbit) -> Observations:
+        pass
+
+
+class FixedRadiusObservationSource(ObservationSource):
+    """A FixedRadiusObservationSource is an ObservationSource that
+    gathers observations within a fixed radius of the test orbit's
+    ephemeris at each exposure time within a collection of exposures.
+
+    """
+
+    def __init__(self, radius: float, all_observations: Observations):
```

Comment:
You've made me a type hinting stickler now. Do `__init__`s normally have a return type?

Reply:
Nice catch! `__init__` always returns `None`, and is assumed to do so. There is a lengthy and boring thread on this at https://github.com/python/mypy/issues/604, resolved in https://github.com/python/mypy/pull/5677.

There's nothing wrong with explicitly adding `-> None` there though.

You might be a bit surprised that it returns `None` and not, say, `Self`. That's because `__new__` is the actual constructor function which returns the new class instance. `__init__` is actually a hook that gets called to initialize the new instance, and its return value is discarded.